File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (toUpper, toLower, isSpace)
import Data.List (dropWhileEnd)

responseFor :: String -> String
responseFor xs
  | isSilence xs'          = "Fine. Be that way!"
  | isForcefulQuestion xs' = "Calm down, I know what I'm doing!"
  | isShouting xs'         = "Whoa, chill out!"
  | isQuestion xs'         = "Sure."
  | otherwise              = "Whatever."
  where xs'                   = trim xs
        trim                  = dropWhile isSpace . dropWhileEnd isSpace
        isForcefulQuestion xs = isShouting xs && isQuestion xs
        isShouting xs         = xs == map toUpper xs
                                && not (map toUpper xs == map toLower xs)
        isSilence             = null
        isQuestion xs         = last xs == '?'

File: fsharp/bob/Bob.fs
module Bob

let isShouting (input: string): bool =
  input = input.ToUpper() && input <> input.ToLower()

let isQuestion (input: string): bool =
  input |> Seq.last = '?'

let isForcefulQuestion (input: string): bool =
  isShouting(input) && isQuestion(input)

let isSilence = Seq.isEmpty

let response (input: string): string =
  let input = input.Trim()
  if isSilence(input) then "Fine. Be that way!"
  elif isForcefulQuestion(input) then "Calm down, I know what I'm doing!"
  elif isShouting(input) then "Whoa, chill out!"
  elif isQuestion(input) then "Sure."
  else "Whatever."
